lab5-scott/lib/array-method.js

```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+const List = module.exports = function() {
+  this.length = 0;
+};
+
+List.prototype.push = function(value) {
+  this[this.length] = value;
+  this.length++;
+};
+
+List.prototype.pop = function(){
+  if (this.length < 1) return;
+  let result = this[this.length - 1];
+  delete this[this.length - 1];
+  this.length--;
+  return result;
+};
+
+List.prototype.map = function(callback){
+  let result = new List();
+  for (var i = 0; i < this.length; i++) {
+    result.push(callback(this[i]));
+  }
+  return result;
+};
+
+List.prototype.filter = function(callback){
+  let result = new List();
+  for (var i = 0; i < this.length; i++) {
+    if (callback(this[i])) {
+      delete this[i];
+    } else{
+      result.push(this[i]);
+    }
```

Comment:
Because filter returns a new list we don't actually have to delete anything from the original list. Also, wouldn't we want to push to the result List if it passes the condition defined in the callback, as opposed to if it fails the condition?